File: server/sources/registry.js
// registry.js — the pluggability core. Every adapter registers here; the route just
// asks collect('incidents'|'cameras', bbox, cfg) and gets a fused, de-duplicated list
// plus an HONEST per-source health array (mirrors the Air domain's link-health ethos —
// a dead feed shows as {ok:false}, it never fails the whole route).
//
// Keyless official feeds are always on. Opt-in "gray" adapters (Citizen / Snap Map /
// PulsePoint / scanner) are OFF unless their env flag is set — they are never a default.

import { bboxFromRadius } from './types.js';
import socrata from './socrata.js';
import arcgis from './arcgis.js';
import nws from './nws.js';
import usgs from './usgs.js';
import caltrans from './caltrans.js';
import nyctmc from './nyctmc.js';
import citizen from './gray/citizen.js';
import pulsepoint from './gray/pulsepoint.js';
import snapmap from './gray/snapmap.js';
import scanner from './gray/scanner.js';
// Phase 2 — keyed-but-free (off until an env key is set) + keyless TfL/FL511.
import tfl from './tfl.js';
import airquality from './airquality.js';
import firms from './firms.js';
import windy from './windy.js';
import wsdot from './wsdot.js';
import open511sf from './open511sf.js';
import events from './events.js';
// Phase 3 — more keyless natural-hazard feeds + opt-in Bluesky social.
import iem from './iem.js';
import eonet from './eonet.js';
import gdacs from './gdacs.js';
import nwps from './nwps.js';
// Real-time California road activity (the live backbone for LA, which publishes no
// real-time police/fire dispatch feed).
import caltransLcs from './caltrans-lcs.js';
import caltransCms from './caltrans-cms.js';
import bluesky from './gray/bluesky.js';

const ADAPTERS = [
  // --- keyless, official / legal-aggregator, DEFAULT ON ---
  ...socrata, ...arcgis, ...nws, ...usgs,           // incidents (+ FL511 cameras via arcgis)
  ...iem, ...eonet, ...gdacs, ...nwps,              // more keyless hazard / natural-event feeds
  ...caltransLcs, ...caltransCms,                    // real-time CA road closures + message signs
  ...caltrans, ...nyctmc, ...tfl,                    // cameras
  // --- Phase 2: keyed-but-free, activate by setting the env key (off otherwise) ---
  ...airquality, ...firms, ...windy, ...wsdot, ...open511sf, ...events,
  // --- opt-in "gray", DEFAULT OFF (never enabled without an explicit flag) ---
  ...citizen, ...pulsepoint, ...snapmap, ...scanner, ...bluesky,
];

// id -> image URL for every camera we've SERVED, so /api/camimg/:id resolves against the
// server's own catalog (never a caller-supplied URL: no open image proxy / SSRF).
export const cameraIndex = new Map();
const CAM_INDEX_MAX = 8000;
function indexCamera(cam) {
  const url = cam.still || cam.stream;
  if (!url) return;
  cameraIndex.delete(cam.id);
  cameraIndex.set(cam.id, url);
  if (cameraIndex.size > CAM_INDEX_MAX) cameraIndex.delete(cameraIndex.keys().next().value);
}

function parseScanner(s) {
  return String(s || '').split(',').map((t) => t.trim()).filter(Boolean).map((tok) => {
    const [shortName, lat, lon, ...label] = tok.split(':');
    return { shortName, lat: Number(lat), lon: Number(lon), label: label.join(':') || shortName };
  }).filter((x) => x.shortName && Number.isFinite(x.lat) && Number.isFinite(x.lon));
}

export function resolveConfig(env = process.env) {
  return {
    socrataToken: env.SOCRATA_APP_TOKEN || null,
    // Phase 2 keyed feeds — free to obtain; each stays OFF until its key is present.
    airnowKey: env.AIRNOW_KEY || null,
    firmsKey: env.FIRMS_MAP_KEY || null,
    windyKey: env.WINDY_KEY || null,
    wsdotKey: env.WSDOT_KEY || null,
    tflKey: env.TFL_APP_KEY || null,               // optional: TfL works keyless at low rate
    five11SfToken: env.FIVE11_SF_TOKEN || null,
    ticketmasterKey: env.TICKETMASTER_KEY || null,
    enableCitizen: env.VANTAGE_ENABLE_CITIZEN === '1',
    enableSnap: env.VANTAGE_ENABLE_SNAPMAP === '1',
    enablePulsepoint: env.VANTAGE_ENABLE_PULSEPOINT === '1',
    pulsepointAgencies: String(env.VANTAGE_PULSEPOINT_AGENCIES || '').split(',').map((x) => x.trim()).filter(Boolean),
    enableScanner: env.VANTAGE_ENABLE_SCANNER === '1',
    scannerSystems: parseScanner(env.VANTAGE_SCANNER_SYSTEMS),
    blueskyQuery: env.VANTAGE_BLUESKY_QUERY || null, // opt-in aggregate social by place term
  };
}

function safeEnabled(a, cfg) {
  try { return a.enabled ? !!a.enabled(cfg) : true; } catch { return false; }
}

// list adapters + their resolved on/off state (for a status endpoint / the UI legend).
export function listAdapters(cfg) {
  return ADAPTERS.map((a) => ({
    id: a.id, category: a.category, kinds: a.kinds || [], optin: !!a.optin,
    keyed: a.keyless === false,
    enabled: safeEnabled(a, cfg), attribution: a.attribution || '', label: a.label || a.id,
  }));
}

// Hard per-adapter wall so one slow feed can't hold the whole fused response (each
// adapter also self-times-out at 8 s in _http.js; this is the belt-and-braces backstop).
const ADAPTER_WALL_MS = 9000;
function withWall(p) {
  return Promise.race([
    Promise.resolve(p),
    new Promise((_, rej) => setTimeout(() => rej(new Error('adapter timeout')), ADAPTER_WALL_MS)),
  ]);
}

export async function collect(category, bbox, cfg) {
  const chosen = ADAPTERS.filter((a) => a.category === category && safeEnabled(a, cfg));
  const settled = await Promise.allSettled(chosen.map((a) => withWall(a.fetch(bbox, cfg))));
  const items = [];
  const sources = [];
  chosen.forEach((a, i) => {
    const r = settled[i];
    if (r.status === 'fulfilled' && Array.isArray(r.value)) {
      const clean = r.value.filter(Boolean);
      items.push(...clean);
      if (category === 'cameras') for (const cam of clean) indexCamera(cam);
      sources.push({ id: a.id, ok: true, count: clean.length, optin: !!a.optin });
    } else {
      sources.push({ id: a.id, ok: false, count: 0, optin: !!a.optin, error: String(r.reason || 'failed').slice(0, 140) });
    }
  });
  // de-dup by stable id (same incident reported by two feeds collapses to one)
  const seen = new Set();
  const deduped = [];
  for (const it of items) { if (!it || seen.has(it.id)) continue; seen.add(it.id); deduped.push(it); }
  return { items: deduped, sources };
}

export { bboxFromRadius };

File: src-tauri/src/proxy/sources/registry.rs
// registry.rs — the native mirror of server/sources/registry.js. Two jobs:
//   1. resolve_config(): read the same env vars Node's resolveConfig() reads, so keyed and
//      opt-in ("gray") sources activate identically on the desktop.
//   2. CATALOG + list_sources(): the descriptor table behind GET /api/sources, key-for-key
//      identical to Node's listAdapters() output {id,category,kinds,optin,keyed,enabled,
//      attribution,label}. The incidents/cameras handlers gate the keyed/gray fetches on the
//      same enabled(cfg) predicates, so with no keys the fused source set matches Node exactly.

use serde_json::{json, Value};

/// One configured scanner system (VANTAGE_SCANNER_SYSTEMS="short:lat:lon:Label,...").
#[derive(Clone)]
pub struct ScannerSys {
    pub short_name: String,
    pub lat: f64,
    pub lon: f64,
    pub label: String,
}

/// Resolved runtime config — mirror of Node's resolveConfig() (registry.js:74).
#[derive(Clone, Default)]
pub struct Config {
    pub socrata_token: Option<String>,
    pub airnow_key: Option<String>,
    pub firms_key: Option<String>,
    pub windy_key: Option<String>,
    pub wsdot_key: Option<String>,
    pub tfl_key: Option<String>,
    pub five11_sf_token: Option<String>,
    pub ticketmaster_key: Option<String>,
    pub enable_citizen: bool,
    pub enable_snap: bool,
    pub enable_pulsepoint: bool,
    pub pulsepoint_agencies: Vec<String>,
    pub enable_scanner: bool,
    pub scanner_systems: Vec<ScannerSys>,
    pub bluesky_query: Option<String>,
}

fn env_opt(name: &str) -> Option<String> {
    std::env::var(name).ok().map(|s| s.trim().to_string()).filter(|s| !s.is_empty())
}
fn env_flag(name: &str) -> bool {
    env_opt(name).as_deref() == Some("1")
}
fn csv(name: &str) -> Vec<String> {
    env_opt(name)
        .map(|s| s.split(',').map(|t| t.trim().to_string()).filter(|t| !t.is_empty()).collect())
        .unwrap_or_default()
}
fn parse_scanner(name: &str) -> Vec<ScannerSys> {
    let mut out = Vec::new();
    for tok in env_opt(name).unwrap_or_default().split(',') {
        let parts: Vec<&str> = tok.split(':').collect();
        if parts.len() < 3 {
            continue;
        }
        let short = parts[0].trim();
        let (lat, lon) = match (parts[1].trim().parse::<f64>(), parts[2].trim().parse::<f64>()) {
            (Ok(a), Ok(o)) if a.is_finite() && o.is_finite() => (a, o),
            _ => continue,
        };
        if short.is_empty() {
            continue;
        }
        let label = if parts.len() > 3 && !parts[3].trim().is_empty() { parts[3..].join(":") } else { short.to_string() };
        out.push(ScannerSys { short_name: short.to_string(), lat, lon, label });
    }
    out
}

pub fn resolve_config() -> Config {
    Config {
        socrata_token: env_opt("SOCRATA_APP_TOKEN"),
        airnow_key: env_opt("AIRNOW_KEY"),
        firms_key: env_opt("FIRMS_MAP_KEY"),
        windy_key: env_opt("WINDY_KEY"),
        wsdot_key: env_opt("WSDOT_KEY"),
        tfl_key: env_opt("TFL_APP_KEY"),
        five11_sf_token: env_opt("FIVE11_SF_TOKEN"),
        ticketmaster_key: env_opt("TICKETMASTER_KEY"),
        enable_citizen: env_flag("VANTAGE_ENABLE_CITIZEN"),
        enable_snap: env_flag("VANTAGE_ENABLE_SNAPMAP"),
        enable_pulsepoint: env_flag("VANTAGE_ENABLE_PULSEPOINT"),
        pulsepoint_agencies: csv("VANTAGE_PULSEPOINT_AGENCIES"),
        enable_scanner: env_flag("VANTAGE_ENABLE_SCANNER"),
        scanner_systems: parse_scanner("VANTAGE_SCANNER_SYSTEMS"),
        bluesky_query: env_opt("VANTAGE_BLUESKY_QUERY"),
    }
}

// --- descriptor catalog (order + fields mirror Node's ADAPTERS / listAdapters) -----------

pub struct Descriptor {
    pub id: &'static str,
    pub category: &'static str,
    pub kinds: &'static [&'static str],
    pub keyed: bool,
    pub optin: bool,
    pub attribution: &'static str,
    pub label: &'static str,
    pub enabled: fn(&Config) -> bool,
}

// enabled predicates (fn items so the table can be a const)
fn on(_: &Config) -> bool { true }
fn e_airnow(c: &Config) -> bool { c.airnow_key.is_some() }
fn e_firms(c: &Config) -> bool { c.firms_key.is_some() }
fn e_windy(c: &Config) -> bool { c.windy_key.is_some() }
fn e_wsdot(c: &Config) -> bool { c.wsdot_key.is_some() }
fn e_511(c: &Config) -> bool { c.five11_sf_token.is_some() }
fn e_tm(c: &Config) -> bool { c.ticketmaster_key.is_some() }
fn e_citizen(c: &Config) -> bool { c.enable_citizen }
fn e_pulsepoint(c: &Config) -> bool { c.enable_pulsepoint && !c.pulsepoint_agencies.is_empty() }
fn e_snap(c: &Config) -> bool { c.enable_snap }
fn e_scanner(c: &Config) -> bool { c.enable_scanner && !c.scanner_systems.is_empty() }
fn e_bsky(c: &Config) -> bool { c.bluesky_query.is_some() }

macro_rules! d {
    ($id:expr, $cat:expr, $kinds:expr, $keyed:expr, $optin:expr, $attr:expr, $label:expr, $en:expr) => {
        Descriptor { id: $id, category: $cat, kinds: &$kinds, keyed: $keyed, optin: $optin, attribution: $attr, label: $label, enabled: $en }
    };
}

pub const CATALOG: &[Descriptor] = &[
    // keyless core (always on) — same order as server/sources/registry.js ADAPTERS
    d!("sea-fire-cad", "incidents", ["fire", "medical"], false, false, "Seattle Fire · data.seattle.gov", "Seattle Fire/EMS dispatch", on),
    d!("sf-pd-cad", "incidents", ["police"], false, false, "SFPD · data.sfgov.org", "SF Police dispatch (real-time)", on),
    d!("sf-fire-cad", "incidents", ["fire", "medical"], false, false, "SF Fire · data.sfgov.org", "SF Fire calls", on),
    d!("chi-311", "incidents", ["civic"], false, false, "City of Chicago · data.cityofchicago.org", "Chicago 311 service requests", on),
    d!("chi-crime", "incidents", ["police"], false, false, "Chicago PD · data.cityofchicago.org", "Chicago crimes (7-day lag)", on),
    d!("nyc-311", "incidents", ["civic"], false, false, "NYC OpenData · data.cityofnewyork.us", "NYC 311 service requests", on),
    d!("cin-cad", "incidents", ["police"], false, false, "Cincinnati PD · data.cincinnati-oh.gov", "Cincinnati PD dispatch", on),
    d!("dc-mpd", "incidents", ["police"], false, false, "DC Metropolitan Police · dcgis.dc.gov", "DC Police incidents", on),
    d!("fl511-cam", "cameras", ["camera"], false, false, "Florida DOT · FL511", "Florida DOT cameras", on),
    d!("nws-alerts", "incidents", ["weather", "hazard"], false, false, "US National Weather Service · api.weather.gov", "NWS active alerts", on),
    d!("usgs-quake", "incidents", ["quake"], false, false, "USGS Earthquake Hazards Program", "USGS earthquakes", on),
    d!("usgs-volcano", "incidents", ["hazard"], false, false, "USGS Volcano Hazards Program", "USGS volcano alerts", on),
    d!("iem-lsr", "incidents", ["weather", "hazard"], false, false, "Iowa Environmental Mesonet · NWS", "NWS storm reports (IEM)", on),
    d!("eonet", "incidents", ["fire-wildland", "hazard", "weather"], false, false, "NASA EONET", "NASA EONET natural events", on),
    d!("gdacs", "incidents", ["hazard", "quake", "weather"], false, false, "GDACS · UN/EC", "GDACS global disasters", on),
    d!("nwps", "incidents", ["hazard"], false, false, "NOAA/NWS NWPS", "NWPS flood gauges", on),
    d!("caltrans-lcs", "incidents", ["traffic"], false, false, "Caltrans CWWP2 (no charge)", "Caltrans lane closures (live)", on),
    d!("caltrans-cms", "incidents", ["traffic"], false, false, "Caltrans CWWP2 (no charge)", "Caltrans message signs (live)", on),
    d!("lapd-calls", "incidents", ["police", "fire", "medical", "traffic"], false, false, "LAPD · data.lacity.org", "LAPD calls for service (~5d lag)", on),
    d!("lapd-news", "incidents", ["police"], false, false, "LAPD Online · lapdonline.org", "LAPD news & alerts", on),
    d!("chp-cad", "incidents", ["traffic", "hazard", "police", "medical"], false, false, "California Highway Patrol · cad.chp.ca.gov", "CHP traffic incidents (live)", on),
    d!("caltrans-cam", "cameras", ["camera"], false, false, "Caltrans CWWP2 (no charge)", "Caltrans CCTV", on),
    d!("alertca-cam", "cameras", ["camera"], false, false, "ALERTCalifornia · UC San Diego", "ALERTCalifornia PTZ cameras", on),
    d!("nyc-dot-cam", "cameras", ["camera"], false, false, "NYC DOT · nyctmc.org", "NYC DOT cameras", on),
    d!("tfl-jamcam", "cameras", ["camera"], false, false, "Powered by TfL Open Data", "TfL JamCams (London)", on),
    // keyed-but-free (OFF until an env key is set)
    d!("airnow", "incidents", ["hazard"], true, false, "US EPA AirNow", "AirNow air quality", e_airnow),
    d!("firms", "incidents", ["fire-wildland"], true, false, "NASA FIRMS (VIIRS S-NPP)", "NASA FIRMS wildfire", e_firms),
    d!("windy-cam", "cameras", ["camera"], true, false, "Windy Webcams", "Windy webcams", e_windy),
    d!("wsdot-cam", "cameras", ["camera"], true, false, "WSDOT", "WSDOT cameras (WA)", e_wsdot),
    d!("wsdot-alerts", "incidents", ["traffic"], true, false, "WSDOT", "WSDOT traffic alerts (WA)", e_wsdot),
    d!("511sfbay", "incidents", ["traffic", "civic"], true, false, "511.org · MTC", "511 SF Bay traffic", e_511),
    d!("ticketmaster", "incidents", ["civic"], true, false, "Ticketmaster Discovery", "Ticketmaster events", e_tm),
    // opt-in "gray" (OFF unless an explicit flag is set)
    d!("citizen", "incidents", ["police", "fire", "medical", "hazard"], false, true, "Citizen (unofficial · place-only)", "Citizen incidents", e_citizen),
    d!("pulsepoint", "incidents", ["fire", "medical"], false, true, "PulsePoint (in-the-clear · place-only)", "PulsePoint fire/EMS", e_pulsepoint),
    d!("snapmap", "incidents", ["social"], false, true, "Snap Map (aggregate place-heat only)", "Snap Map activity", e_snap),
    d!("scanner", "incidents", ["police"], false, true, "OpenMHz scanner (aggregate activity)", "Scanner activity", e_scanner),
    d!("bluesky", "incidents", ["social"], false, true, "Bluesky (aggregate place-heat)", "Bluesky social chatter", e_bsky),
];

/// GET /api/sources body — mirror of Node listAdapters(cfg).
pub fn list_sources(cfg: &Config) -> Vec<Value> {
    CATALOG
        .iter()
        .map(|d| {
            json!({
                "id": d.id,
                "category": d.category,
                "kinds": d.kinds,
                "optin": d.optin,
                "keyed": d.keyed,
                "enabled": (d.enabled)(cfg),
                "attribution": d.attribution,
                "label": d.label,
            })
        })
        .collect()
}
